File: BinaryTreeVerticalOrderTraversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def verticalOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        if not root: return []
        rlist = []#vertical place of[0,1,2,3,...]
        llist = []#vertical place of[-1,-2,-3...]
        q = collections.deque()
        q.append((0,root))
        while q:
            posi, node = q.popleft()
            self.helper(posi,node,rlist,llist)
            if node.left:
                q.append((posi-1,node.left))
            if node.right:
                q.append((posi+1,node.right))
        llist.reverse()
        return llist+rlist
        
    def helper(self,posi,node,rlist,llist):
        if not node: return
        if posi >=0: 
            if posi < len(rlist):#if position is within the span of list,take sublist out and insert n
                tmpl = rlist.pop(posi)
                tmpl.append(node.val)
                rlist.insert(posi,tmpl)
            else:#if position is out of bound of list, append new sublist to list
                tmpl = [node.val]
                rlist.append(tmpl)
        else:
            posi = -posi
            if posi-1 < len(llist):
                tmpl = llist.pop(posi-1)
                tmpl.append(node.val)
                llist.insert(posi-1,tmpl)
            else:
                tmpl = [node.val]
                llist.append(tmpl)
            posi = -posi
        

File: BinaryTreeVerticalOrderTraversal.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    var result = mutableListOf<List<Int>>()
    var colMap = mutableMapOf<Int, MutableList<Pair<Int,Int>>>()
    fun verticalOrder(root: TreeNode?): List<List<Int>> {
        if (root == null) return result
        helper(0, 0, root)
        var keys = colMap.keys
        val start = keys.min()
        val end = keys.max()
        for (idx in start .. end) {
            val tmpPairs: MutableList<Pair<Int,Int>> = colMap[idx]?: mutableListOf<Pair<Int,Int>>()
            val tmpList: List<Int> = tmpPairs.sortedBy{it.first}.map{it.second}
            result += tmpList
        }
        return result
    }

    fun helper(row: Int, colNumber: Int, node: TreeNode?) {
        if (node == null) return
        var currList = colMap[colNumber]?: mutableListOf<Pair<Int,Int>>()
        currList += Pair(row, node!!.`val`)
        colMap.put(colNumber, currList)
        if (node!!.left != null) helper(row + 1, colNumber - 1, node!!.left)
        if (node!!.right != null) helper(row + 1, colNumber + 1, node!!.right)
    }
}
